### PR TITLE
limit compression pointers to 14 bits

### PR DESCRIPTION
**Pull request by Habbie**
### Short description
In a very specific edge case, we would get compression pointers above the 0-0x3fff range, and truncate those to a very low number, causing clients to throw parse errors.

I have added a fix from @pieterlexis where DNSPacketWriter did not honour `compress=false` for the content of any type (like NS) that is permitted to compress by RFC.

### Checklist

I have:
- [ ] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

File: pdns/dnswriter.cc [hunk 1]
```diff
@@ -325,7 +325,7 @@ void DNSPacketWriter::xfrName(const DNSName& name, bool compress, bool)
 
   uint16_t li=0;
   uint16_t matchlen=0;
-  if(compress && (li=lookupName(name, &matchlen))) {
+  if(compress && (li=lookupName(name, &matchlen)) && li < 16384) {
```
Comment (rgacogne) [inline pdns/dnswriter.cc]: I wonder if that check should live in `DNSPacketWriter::lookupName()` instead so we can try to find an existing name with less matching labels?

Comment (Habbie) [reply pdns/dnswriter.cc]: ah yes

Comment (Habbie) [reply pdns/dnswriter.cc]: pushed

**Comment by Habbie** [comment]
This PR contains two commits. i have verified that either commit, independently, fixes the bug. The second commit is an optimisation suggested by @rgacogne.

**Comment by Habbie** [comment]
After discussion with @omoerbeek and @rgacogne, I have pushed a third commit that also, independently, fixes the issue. I think we want to ship the first and third commits.

File: pdns/dnswriter.cc [hunk 2]
```diff
@@ -266,6 +266,8 @@ uint16_t DNSPacketWriter::lookupName(const DNSName& name, uint16_t* matchLen)
         }
         if(!c)
           break;
+        auto offset = iter - d_content.cbegin();
+        if (offset >= 16384) break; // compression pointers cannot point here
         pvect.push_back(iter - d_content.cbegin());
```
Comment (rgacogne) [inline pdns/dnswriter.cc]:
```suggestion
        pvect.push_back(offset);
```

Comment (Habbie) [reply pdns/dnswriter.cc]: duh

Comment (Habbie) [reply pdns/dnswriter.cc]: pushed

**Comment by omoerbeek** [comment]
Would it be posisble to transform you test code in a unit test?

From close source reading the suggested fix looks right (commits 1+3+4). Although 1 is not strictly necessary. Existing unittests also succeed for all three products here.

**Comment by Habbie** [comment]
> Would it be possible to transform you test code into a unit test?

Probably, and I think it's a good idea. It might be the kind of test that 'breaks' whenever we change something in this area, though.

> Although 1 is not strictly necessary.

Correct, but it has no downsides and thus is a nice failsafe.

**Comment by Habbie** [comment]
I pushed one extremely specific test.

File: pdns/dnswriter.cc [hunk 3]
```diff
@@ -266,7 +267,9 @@ uint16_t DNSPacketWriter::lookupName(const DNSName& name, uint16_t* matchLen)
         }
         if(!c)
           break;
-        pvect.push_back(iter - d_content.cbegin());
+        auto offset = iter - d_content.cbegin();
+        if (offset >= 16384) break; // compression pointers cannot point here
```
Comment (pieterlexis) [inline pdns/dnswriter.cc]: It might make sense to
```cpp
#define MAX_COMPRESSION_POINTER 16384
```

and use it everywhere

**Comment by Habbie** [comment]
Squashed, ready for final review I hope!

**Pull request merged**